proxy_ring_buffer.go

```diff
@@ -0,0 +1,278 @@
+// The initial value of the head is 0 (1 more than tail).
+func (rb *Ringbuffer) HeadSequence(ctx context.Context) (int64, error) {
+	request := codec.EncodeRingbufferHeadSequenceRequest(rb.name)
+	response, err := rb.invokeOnPartition(ctx, request, rb.partitionID)
+	if err != nil {
+		return ReadResultSetSequenceUnavailable, err
+	}
+	return codec.DecodeRingbufferHeadSequenceResponse(response), nil
+}
+
+// RemainingCapacity
+// Returns the remaining capacity of the Ringbuffer. The returned value could be stale as soon as it is returned.
+// If ttl is not set, the remaining capacity will always be the capacity.
+func (rb *Ringbuffer) RemainingCapacity(ctx context.Context) (int64, error) {
+	request := codec.EncodeRingbufferRemainingCapacityRequest(rb.name)
+	response, err := rb.invokeOnPartition(ctx, request, rb.partitionID)
+	if err != nil {
+		return ReadResultSetSequenceUnavailable, err
+	}
+	return codec.DecodeRingbufferRemainingCapacityResponse(response), nil
+}
+
+// ReadMany
+// Reads a batch of items from the Ringbuffer. If the number of available items after the first read item is smaller
+// than the maxCount, these items are returned. So it could be the number of items read is smaller than the maxCount.
+// If there are fewer items available than minCount, then this call blacks. Reading a batch of items is likely to
+// perform better because less overhead is involved. A filter can be provided to only select items that need to be read.
+// If the filter is null, all items are read. If the filter is not null, only items where the filter function returns
+// true are returned. Using filters is a good way to prevent getting items that are of no value to the receiver.
+// This reduces the amount of IO and the number of operations being executed, and can result in a significant performance improvement.
+func (rb *Ringbuffer) ReadMany(ctx context.Context, startSequence int64, minCount int32, maxCount int32, filter interface{}) (ReadResultSet, error) {
+	if filter != nil {
+		return ReadResultSet{}, errors.New("filter functions are not yet supported in th Go client, please use 'nil' for now")
+	}
+	request := codec.EncodeRingbufferReadManyRequest(rb.name, startSequence, minCount, maxCount, nil)
+	response, err := rb.invokeOnPartition(ctx, request, rb.partitionID)
+	if err != nil {
+		return ReadResultSet{}, err
+	}
+	readCount, items, itemSeqs, nextSeq := codec.DecodeRingbufferReadManyResponse(response)
+	return ReadResultSet{
+		rb:        rb,
+		readCount: readCount,
+		items:     items,
+		itemSeqs:  itemSeqs,
+		nextSeq:   nextSeq,
+	}, nil
+}
+
+// ReadCount Number of items that have been read before filtering.
+func (rrs *ReadResultSet) ReadCount() int32 {
+	return rrs.readCount
+}
+
+// Get one item from List of items that have been read.
+func (rrs *ReadResultSet) Get(index int) (interface{}, error) {
+	if err := rrs.validateIndexInRange(index); err != nil {
+		return nil, err
+	}
+	return rrs.rb.convertToObject(rrs.items[index])
```

Comment:
In TDD we discussed lazy deserialization and we don't want any lazy deserialization for now. So, I guess you can deserialize all the items in `ReadMany()` cc @yuce @utku-caglayan

Reply:
I'm sorry, but I feel lost, in what was discussed in "TDD".
Is this already decided (and I shall change it) OR an open question?

Reply:
Yeah it is decided. You can deserialize all items when you create the ReadResultSet and pass them to ReadResultSet. Then Get() will return the item directly.

Reply:
Done.
See 6e4837544829edc4dd39c191be46b65cf76cfbea